jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/FrameFlusher.java

```diff
@@ -55,60 +55,71 @@ public boolean isControlFrame()
-    private final LongAdder bytesOut = new LongAdder();
-    private final ByteBufferPool bufferPool;
-    private final EndPoint endPoint;
-    private final int bufferSize;
-    private final Generator generator;
-    private final int maxGather;
-    private final Deque<Entry> queue = new ArrayDeque<>();
-    private final List<ByteBuffer> buffers;
-    private final CyclicTimeout cyclicTimeout;
-    private final List<Entry> entries;
-    private final List<Entry> previousEntries;
-    private final List<Entry> failedEntries;
-
-    private List<RetainableByteBuffer> releasableBuffers = new ArrayList<>();
-    private RetainableByteBuffer batchBuffer;
-    private boolean canEnqueue = true;
-    private boolean flushed = true;
-    private Throwable closedCause;
-    private long idleTimeout;
-    private boolean useDirectByteBuffers;
+
+    private final AutoLock _lock = new AutoLock();
+    private final LongAdder _messagesOut = new LongAdder();
+    private final LongAdder _bytesOut = new LongAdder();
+    private final ByteBufferPool _bufferPool;
+    private final EndPoint _endPoint;
+    private final int _bufferSize;
+    private final Generator _generator;
+    private final int _maxGather;
+    private final Deque<Entry> _queue = new ArrayDeque<>();
+    private final CyclicTimeouts<Entry> _cyclicTimeouts;
+    private final List<Entry> _entries;
+    private final List<RetainableByteBuffer> _releasableBuffers = new ArrayList<>();
+
+    private RetainableByteBuffer _batchBuffer;
+    private boolean _canEnqueue = true;
+    private Throwable _closedCause;
+    private long _frameTimeout;
+    private long _messageTimeout;
+    private boolean _useDirectByteBuffers;
+    private long _currentMessageTimeout;
 
     public FrameFlusher(ByteBufferPool bufferPool, Scheduler scheduler, Generator generator, EndPoint endPoint, int bufferSize, int maxGather)
     {
-        this.bufferPool = bufferPool;
-        this.endPoint = endPoint;
-        this.bufferSize = bufferSize;
-        this.generator = Objects.requireNonNull(generator);
-        this.maxGather = maxGather;
-        this.entries = new ArrayList<>(maxGather);
-        this.previousEntries = new ArrayList<>(maxGather);
-        this.failedEntries = new ArrayList<>(maxGather);
-        this.buffers = new ArrayList<>((maxGather * 2) + 1);
-        this.cyclicTimeout = new CyclicTimeout(scheduler)
+        _bufferPool = bufferPool;
+        _endPoint = endPoint;
+        _bufferSize = bufferSize;
+        _generator = Objects.requireNonNull(generator);
+        _maxGather = maxGather;
+        _entries = new ArrayList<>(maxGather);
+        _cyclicTimeouts = new CyclicTimeouts<>(scheduler)
         {
             @Override
-            public void onTimeoutExpired()
+            protected boolean onExpired(Entry expirable)
+            {
+                abort(new WebSocketWriteTimeoutException("FrameFlusher Frame Write Timeout"));
```

Comment:
comment that this called with the lock held.   The abort method also grabs the lock again, which is kind of wrong.   Perhaps you need an `abortLocked(...)` method that test that the lock is already held and call that directly here.   The normal `abort(...)` method can just grab the lock call `abortLocked(...)`....

Hmmm but that still doesn't work as the current method releases the lock before calling `onCompleteFailure`, so this version will call that with the lock held!